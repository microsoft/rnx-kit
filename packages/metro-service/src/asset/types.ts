export type PackagerAsset = {
  httpServerLocation: string;
  name: string;
  type: string;
};
