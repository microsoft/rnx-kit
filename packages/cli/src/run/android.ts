import type { Config } from "@react-native-community/cli-types";
import * as path from "node:path";
import ora from "ora";
import { buildAndroid } from "../build/android";
import type { AndroidInputParams } from "../build/types";

export async function runAndroid(
  config: Config,
  buildParams: AndroidInputParams
) {
  const logger = ora();

  const projectDir = await buildAndroid(config, buildParams, logger);
  if (typeof projectDir !== "string") {
    return;
  }

  logger.start("Preparing to launch app");

  const { findOutputFile, getPackageName, install, selectDevice, start } =
    await import("@rnx-kit/tools-android");

  const { configuration = "Debug" } = buildParams;
  const apks = findOutputFile(projectDir, configuration);
  if (apks.length === 0) {
    logger.fail("Failed to find the APK that was just built");
    process.exitCode = 1;
    return;
  }

  if (apks.length > 1) {
    const currentStatus = logger.text;
    const choices = apks.map((p) => path.basename(p)).join(", ");
    logger.info(`Multiple APKs were found; picking the first one: ${choices}`);
    logger.info("If this is wrong, remove the others and try again");
    logger.start(currentStatus);
  }

  const apk = apks[0];
  const info = getPackageName(apk);
  if (info instanceof Error) {
    logger.fail(info.message);
    process.exitCode = 1;
    return;
  }

  const device = await selectDevice(buildParams.device, logger);
  if (!device) {
    logger.fail("Failed to launch app: Could not find an appropriate device");
    process.exitCode = 1;
    return;
  }

  logger.start(`Installing ${apk}`);

  const { packageName, activityName } = info;
  const error = await install(device, apk, packageName);
  if (error) {
    logger.fail(error.message);
    process.exitCode = 1;
    return;
  }

  logger.text = `Starting ${packageName}`;
  await start(device, packageName, activityName);

  logger.succeed(`Started ${packageName}`);
}
