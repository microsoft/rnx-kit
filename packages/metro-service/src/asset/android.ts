import { isApproximatelyEqual } from "@rnx-kit/tools-language/math";
import path from "path";
import type { PackagerAsset } from "./types";

export function getAndroidAssetSuffix(
  asset: PackagerAsset,
  scale: number
): string {
  const tolerance = 0.01;
  if (isApproximatelyEqual(scale, 0.75, tolerance)) {
    return "ldpi";
  } else if (isApproximatelyEqual(scale, 1, tolerance)) {
    return "mdpi";
  } else if (isApproximatelyEqual(scale, 1.5, tolerance)) {
    return "hdpi";
  } else if (isApproximatelyEqual(scale, 2, tolerance)) {
    return "xhdpi";
  } else if (isApproximatelyEqual(scale, 3, tolerance)) {
    return "xxhdpi";
  } else if (isApproximatelyEqual(scale, 4, tolerance)) {
    return "xxxhdpi";
  }

  throw new Error(
    `Don't know which android drawable suffix to use for asset: ${JSON.stringify(
      asset
    )}`
  );
}

// See https://developer.android.com/guide/topics/resources/drawable-resource.html
const drawableFileTypes = new Set<string>([
  "gif",
  "jpeg",
  "jpg",
  "png",
  "webp",
  "xml",
]);

function getAndroidResourceFolderName(
  asset: PackagerAsset,
  scale: number
): string {
  if (!drawableFileTypes.has(asset.type)) {
    return "raw";
  }
  return `drawable-${getAndroidAssetSuffix(asset, scale)}`;
}

function getBasePath(asset: PackagerAsset): string {
  let basePath = asset.httpServerLocation;
  if (basePath[0] === "/") {
    basePath = basePath.substr(1);
  }
  return basePath;
}

function getAndroidResourceIdentifier(asset: PackagerAsset): string {
  const folderPath = getBasePath(asset);
  return `${folderPath}/${asset.name}`
    .toLowerCase()
    .replace(/\//g, "_") // Encode folder structure in file name
    .replace(/([^a-z0-9_])/g, "") // Remove illegal chars
    .replace(/^assets_/, ""); // Remove "assets_" prefix
}

export function getAssetDestPathAndroid(
  asset: PackagerAsset,
  scale: number
): string {
  const androidFolder = getAndroidResourceFolderName(asset, scale);
  const fileName = getAndroidResourceIdentifier(asset);
  return path.join(androidFolder, `${fileName}.${asset.type}`);
}
