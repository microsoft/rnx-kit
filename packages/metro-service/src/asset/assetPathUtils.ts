// https://github.com/react-native-community/cli/blob/716555851b442a83a1bf5e0db27b6226318c9a69/packages/cli-plugin-metro/src/commands/bundle/assetPathUtils.ts

import type { PackagerAsset } from "./types";

function getBasePath(asset: PackagerAsset): string {
  let basePath = asset.httpServerLocation;
  if (basePath[0] === "/") {
    basePath = basePath.substring(1);
  }
  return basePath;
}

export function getResourceIdentifier(asset: PackagerAsset): string {
  const folderPath = getBasePath(asset);
  return `${folderPath}/${asset.name}`
    .toLowerCase()
    .replace(/\//g, "_") // Encode folder structure in file name
    .replace(/([^a-z0-9_])/g, "") // Remove illegal chars
    .replace(/^assets_/, ""); // Remove "assets_" prefix
}
