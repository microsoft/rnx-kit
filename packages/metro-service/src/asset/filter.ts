const ALLOWED_SCALES: { [key: string]: number[] } = {
  ios: [1, 2, 3],
};

export function filterPlatformAssetScales(
  platform: string,
  scales: readonly number[]
): readonly number[] {
  const allowlist: number[] = ALLOWED_SCALES[platform];
  if (!allowlist) {
    return scales;
  }
  const result = scales.filter((scale) => allowlist.indexOf(scale) > -1);
  if (result.length === 0 && scales.length > 0) {
    // No matching scale found, but there are some available. Ideally we don't
    // want to be in this situation and should throw, but for now as a fallback
    // let's just use the closest larger image
    const maxScale = allowlist[allowlist.length - 1];
    for (const scale of scales) {
      if (scale > maxScale) {
        result.push(scale);
        break;
      }
    }

    // There is no larger scales available, use the largest we have
    if (result.length === 0) {
      result.push(scales[scales.length - 1]);
    }
  }
  return result;
}
