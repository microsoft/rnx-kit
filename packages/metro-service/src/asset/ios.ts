import path from "path";
import type { PackagerAsset } from "./types";

export function getAssetDestPathIOS(
  asset: PackagerAsset,
  scale: number
): string {
  const suffix = scale === 1 ? "" : `@${scale}x`;
  const fileName = `${asset.name + suffix}.${asset.type}`;
  return path.join(
    // Assets can have relative paths outside of the project root.
    // Replace `../` with `_` to make sure they don't end up outside of
    // the expected assets directory.
    asset.httpServerLocation.substr(1).replace(/\.\.\//g, "_"),
    fileName
  );
}
