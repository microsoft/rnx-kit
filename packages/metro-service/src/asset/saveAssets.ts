import { findCommunityCliPluginPath } from "@rnx-kit/tools-react-native/cli";
import * as path from "node:path";
import { saveAssetsAndroid } from "./android";
import { saveAssetsDefault } from "./default";
import { saveAssetsIOS } from "./ios";
import type { SaveAssetsPlugin } from "./types";

// Eventually this will be part of the rn config, but we require it on older rn
// versions for win32 and the cli doesn't allow extra config properties.
// See https://github.com/react-native-community/cli/pull/2002
export function getSaveAssetsPlugin(
  platform: string,
  projectRoot: string
): SaveAssetsPlugin {
  if (platform === "win32") {
    try {
      const saveAssetsPlugin = require.resolve(
        "@office-iss/react-native-win32/saveAssetPlugin",
        { paths: [projectRoot] }
      );
      return require(saveAssetsPlugin);
    } catch (_) {
      /* empty */
    }
  }

  // Use `@react-native/community-cli-plugin` when possible
  const pluginPath = findCommunityCliPluginPath(projectRoot);
  if (pluginPath) {
    const { default: saveAssets } = require(
      path.join(pluginPath, "dist", "commands", "bundle", "saveAssets.js")
    );
    return saveAssets;
  }

  switch (platform) {
    case "ios":
      return saveAssetsIOS;
    case "android":
      return saveAssetsAndroid;
    default:
      return saveAssetsDefault;
  }
}
