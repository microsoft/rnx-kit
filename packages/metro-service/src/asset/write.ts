import * as fs from "fs";
import type { AssetData } from "metro";
import * as path from "path";
import { getAssetDestPathAndroid } from "./android";
import { filterPlatformAssetScales } from "./filter";
import { getAssetDestPathIOS } from "./ios";

function copy(src: string, dest: string): void {
  const destDir = path.dirname(dest);
  fs.mkdirSync(destDir, { recursive: true, mode: 0o755 });
  fs.copyFileSync(src, dest);
}

function copyAll(filesToCopy: Record<string, string>): void {
  const queue = Object.keys(filesToCopy);
  if (queue.length === 0) {
    return;
  }

  console.info(`Copying ${queue.length} asset files`);
  for (const src of queue) {
    copy(src, filesToCopy[src]);
  }
}

export async function saveAssets(
  assets: readonly AssetData[],
  platform: string,
  assetsDest: string | undefined
): Promise<void> {
  if (!assetsDest) {
    console.warn("Assets destination folder is not set, skipping...");
    return Promise.resolve();
  }

  const getAssetDestPath =
    platform === "android" ? getAssetDestPathAndroid : getAssetDestPathIOS;

  const filesToCopy: Record<string, string> = Object.create(null); // Map src -> dest
  assets.forEach((asset) => {
    const validScales = new Set(
      filterPlatformAssetScales(platform, asset.scales)
    );
    asset.scales.forEach((scale, idx) => {
      if (!validScales.has(scale)) {
        return;
      }
      const src = asset.files[idx];
      const dest = path.join(assetsDest, getAssetDestPath(asset, scale));
      filesToCopy[src] = dest;
    });
  });

  copyAll(filesToCopy);
  return Promise.resolve();
}
