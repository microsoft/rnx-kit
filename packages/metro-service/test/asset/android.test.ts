import type { PackagerAsset } from "../../src/asset/types";
import { getAndroidAssetSuffix } from "../../src/asset/android";

describe("MetroService", () => {
  const asset: PackagerAsset = {
    httpServerLocation: "https://my-server.com/assets",
    name: "logo",
    type: "jpg",
  };

  test("getAndroidAssetSuffix returns ldpi for scale 0.75", () => {
    expect(getAndroidAssetSuffix(asset, 0.75)).toEqual("ldpi");
  });

  test("getAndroidAssetSuffix returns mdpi for scale 1", () => {
    expect(getAndroidAssetSuffix(asset, 1)).toEqual("mdpi");
  });

  test("getAndroidAssetSuffix returns hdpi for scale 1.5", () => {
    expect(getAndroidAssetSuffix(asset, 1.5)).toEqual("hdpi");
  });

  test("getAndroidAssetSuffix returns xhdpi for scale 2", () => {
    expect(getAndroidAssetSuffix(asset, 2)).toEqual("xhdpi");
  });

  test("getAndroidAssetSuffix returns xxhdpi for scale 3", () => {
    expect(getAndroidAssetSuffix(asset, 3)).toEqual("xxhdpi");
  });

  test("getAndroidAssetSuffix returns xxxhdpi for scale 4", () => {
    expect(getAndroidAssetSuffix(asset, 4)).toEqual("xxxhdpi");
  });

  test("getAndroidAssetSuffix returns ldpi for scale 0.741", () => {
    expect(getAndroidAssetSuffix(asset, 0.741)).toEqual("ldpi");
  });

  test("getAndroidAssetSuffix returns ldpi for scale 0.759", () => {
    expect(getAndroidAssetSuffix(asset, 0.759)).toEqual("ldpi");
  });

  test("getAndroidAssetSuffix throw for invalid scale 0.70", () => {
    expect(() =>
      getAndroidAssetSuffix(asset, 0.7)
    ).toThrowErrorMatchingSnapshot();
  });

  test("getAndroidAssetSuffix throw for invalid scale 1.23", () => {
    expect(() =>
      getAndroidAssetSuffix(asset, 1.23)
    ).toThrowErrorMatchingSnapshot();
  });
});
