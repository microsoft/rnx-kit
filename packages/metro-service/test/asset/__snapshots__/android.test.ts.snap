// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetroService getAndroidAssetSuffix throw for invalid scale 0.70 1`] = `"Don't know which android drawable suffix to use for asset: {\\"httpServerLocation\\":\\"https://my-server.com/assets\\",\\"name\\":\\"logo\\",\\"type\\":\\"jpg\\"}"`;

exports[`MetroService getAndroidAssetSuffix throw for invalid scale 1.23 1`] = `"Don't know which android drawable suffix to use for asset: {\\"httpServerLocation\\":\\"https://my-server.com/assets\\",\\"name\\":\\"logo\\",\\"type\\":\\"jpg\\"}"`;
